grails-app/domain/org/pih/warehouse/core/Location.groovy

```diff
@@ -66,7 +66,13 @@ class Location implements Comparable<Location>, java.io.Serializable {
     static mappedBy = [locations: "parentLocation"]
 
     static constraints = {
-        name(nullable: false, blank: false, maxSize: 255, unique: 'parentLocation')
+        name(nullable: false, blank: false, maxSize: 255, unique: 'parentLocation', validator: { name, obj ->
+            Boolean isLocationDuplicated = obj.parentLocation ?
+                    Location.findAllByNameAndParentLocation(name, obj.parentLocation).size() > 1 :
+                    Location.findAllByNameAndParentLocationIsNull(name).size() > 1
```

Comment:
Since our validators are getting complicated we should move them to a static method, closure, or named query that can be passed in as a validator. 

Something like this
```
    static constraints = {
        name(nullable: false, blank: false, maxSize: 255, unique: 'parentLocation', validator: uniqueNameValidator)
        ...
    }

    static uniqueNameValidator = { String value, Location obj ->
        Boolean isNameDuplicated = obj.parentLocation ?
                    Location.findAllByNameAndParentLocation(name, obj.parentLocation).size() > 1 :
                    Location.findAllByNameAndParentLocationIsNull(name).size() > 1

            // not sure if validator.unique is correct so check the docs to see what the default unique validation error code is 
            return isNameDuplicated ? ['validator.unique'] : true
    }
```
NOTE: we should do the same for the other two properties that use the custom validator, but leave that for later. We can use this as an opportunity to create a reference implementation.

Reply:
This is getting pretty complicated, so we should consider writing some unit tests to show all of the different validations. Not something for @alannadolny to do, but maybe @awalkowiak or me @jmiranda.